src/fmllauncher/java/net/minecraftforge/fml/loading/TracingPrintStream.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2019.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.fml.loading;
+
+import org.apache.logging.log4j.Logger;
+
+import java.io.PrintStream;
+
+/**
+ * PrintStream which redirects it's output to a given logger.
+ *
+ */
+public class TracingPrintStream extends PrintStream
+{
+
+    private static final int BASE_DEPTH = 4;
+    private final Logger logger;
+
+    public TracingPrintStream(Logger logger, PrintStream original)
+    {
+        super(original);
+        this.logger = logger;
+    }
+
+    private void log(String s)
+    {
+        logger.info("{}{}", getPrefix(), s);
+    }
+
+    private static String getPrefix()
+    {
+        StackTraceElement[] elems = Thread.currentThread().getStackTrace();
+        StackTraceElement elem = elems[BASE_DEPTH]; // The caller is always at BASE_DEPTH, including this call.
```

Comment:
This is just ASKING for index out of bounds. So protect yourself in here.

Reply:
This should not be possible the way this method is called. I can add some protection though...

Reply:
"This should not be possible" This is just asking for it to be possible.